csi.proto

```diff
@@ -1912,3 +1930,139 @@ message GetVolumeGroupSnapshotResponse {
   // This field is REQUIRED
   VolumeGroupSnapshot group_snapshot = 1;
 }
+
+// BlockMetadata specifies a data range.
+message BlockMetadata {
+  // This is the 0 based byte position in the volume or snapshot,
+  // measured from the start of the object.
+  // This field is REQUIRED.
+  uint64 byte_offset = 1;
+
+  // This is the size of the data range.
+  // This field is REQUIRED.
+  uint64 size_bytes = 2;
+}
+
+enum BlockMetadataType {
+  // The FIXED_LENGTH value indicates that data ranges are
+  // returned in fixed size blocks.
+  FIXED_LENGTH=0;
+
+  // The VARIABLE_LENGTH value indicates that data ranges
+  // are returned in potentially variable sized extents.
+  VARIABLE_LENGTH=1;
+}
+
+// The GetAllocatedRequest message is used to solicit metadata on the
+// allocated blocks of a snapshot: i.e. this identifies the data ranges
+// that have valid data as they were the target of some previous write
+// operation on the volume.
+message GetAllocatedRequest {
+  // This is the identifier of the snapshot.
+  // This field is REQUIRED.
+  string snapshot_id = 1;
+
+  // This indicates the zero based starting byte position in the volume
+  // snapshot from which the result should be computed. It is intended
+  // to be used to continue a previously interrupted call.
+  // The plugins may round down this offset to the nearest alignment
+  // boundary based on the BlockMetadataType used.
+  // This field is REQUIRED.
+  uint64 starting_offset = 2;
+
+  // This is an optional parameter, and if non-zero it specifies the
+  // maximum number of tuples to be returned in each
+  // GetAllocatedResponse message returned by the RPC stream.
+  // The plugin will determine an appropriate value if 0, and is
+  // always free to send less than the requested value.
+  // This field is OPTIONAL.
+  uint32 max_results = 3;
+
+  // Secrets required by plugin to complete the request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 4 [(csi_secret) = true];
+}
+
+// GetAllocatedResponse messages are returned in gRPC stream.
+// Cumulatively, they provide information on the allocated data
+// ranges in the snapshot.
+message GetAllocatedResponse {
+  // This specifies the style used in the BlockMetadata sequence.
+  // This value must be the same in all such messages returned by
+  // the stream.
+  // This field is REQUIRED.
+  BlockMetadataType block_metadata_type = 1;
+
+  // This returns the logical size of the underlying volume.
+  // This value must be the same in all such messages returned by
+  // the stream.
+  // This field is REQUIRED.
+  uint64 volume_size_bytes = 2;
+
+  // This is list of data range tuples. The number of entries
+  // in this list is controlled by the max_results value in the
+  // GetAllocatedRequest message.
+  // This field is OPTIONAL.
+  repeated BlockMetadata block_metadata = 3;
```

Comment:
```suggestion
  repeated BlockMetadata block_metadata = 3;
  
  // Indicates there are no more allocated blocks in the list.
  // This field is REQUIRED.
  bool end_of_list = 4;
```

Reply:
As I understand it, a gRPC stream returns an EOF on proper termination and an error otherwise. Would that not suffice?

Reply:
My concern here is when max_results is specified, and you get an EOF, how do you know if you need to call it one more time? Or even if max_results is set to zero, can the SP decide to send less than all the results in one stream?

During my review I was going to add like a whole paragraph explaining the relationship between the max_results field and the number of elements in the block_metadata field, and I decided it was too confusing and it would be better to make it explicit.

The core of the problem is that we've introduced BOTH pagination and streaming and this means the client can't rely on the stream ending to indicate that there are no more pages.

Reply:
> The core of the problem is that we've introduced BOTH pagination and streaming and this means the client can't rely on the stream ending to indicate that there are no more pages.

Hmm ... I'm not sure I understand your confusion! Perhaps we can use this [snippet of Go client prototype code](https://github.com/PrasadG193/external-snapshot-metadata/blob/e9d364a93b77d5a2d00193f164c58c9a676491e3/cmd/client/main.go#L217) (written by @PrasadG193) to dissect the problem?
```
	c.initGRPCClient(snapMetaSvc.Spec.CACert, snapMetaSvc.Spec.Address, saToken, snapNamespace)
	stream, err := c.client.GetDelta(ctx, &pgrpc.GetDeltaRequest{
		BaseSnapshotId:   snapNamespace + "/" + baseVolumeSnapshot,
		TargetSnapshotId: snapNamespace + "/" + targetVolumeSnapshot,
		StartingOffset:   0,
		MaxResults:       uint32(256),
	})
	if err != nil {
		return err
	}
	done := make(chan bool)
	fmt.Println("Resp received:")
	go func() {
		for {
			resp, err := stream.Recv()
			if err == io.EOF {
				done <- true //means stream is finished
				return
			}
			if err != nil {
				log.Fatalf("cannot receive %v", err)
			}
			respJson, _ := json.Marshal(resp)
			fmt.Println(string(respJson))
		}
	}()

	<-done //we will wait until all response is received
```
The example above doesn't really illustrate processing of the multiple entries in the returned `block_metadata` slice, as the response is simply dumped as a string in JSON format, but it is clear that instead of the `fmt.Println` a real client would process the `block_metadata` slice ("page") in the `resp`. The stream processing loop itself is exited only on a non-nil `err` value from `stream.Recv()`, where `io.EOF` is not really an error but the formal proper end-of-stream indicator.

The corresponding [server (sidecar) side logic for this](https://github.com/PrasadG193/external-snapshot-metadata/blob/e9d364a93b77d5a2d00193f164c58c9a676491e3/cmd/server/main.go#L202) primarily consists of a loop doing this:
```
func (s *Server) GetDelta(req *pgrpc.GetDeltaRequest, cbtClientStream pgrpc.SnapshotMetadata_GetDeltaServer) error {
	... // call the SP service after assembling the parameters
	done := make(chan bool)
	go func() {
		for {
			... // receive resp from the SP service
			log.Print("Received response from csi driver, proxying to client")
			if err := cbtClientStream.Send(resp); err != nil {
				log.Printf(fmt.Sprintf("cannot send %v", err))
				return
			}
		}
	}()
	<-done //we will wait until all response is received
	return nil
}
```
(The prototype sidecar has very little error handling)

Reply:
Yes, it appears to be the idiomatic way in Go.

Reply:
I don't hate this way of doing it, but even if it's idiomatic in Go, I'm not sure we use errors in this way anywhere else in CSI. I just want to highlight that a boolean flag in the return message would achieve the same effect as using an error to signal the end of iteration.

Reply:
Why not both? Conventionally, client side should keep the connection alive until `io.EOF`, to [avoid leaks](https://pkg.go.dev/google.golang.org/grpc#ClientConn.NewStream). The `end_of_list` is a confirmation from the server to the client that everything has been sent. If there is an `io.EOF` but `end_of_list` is `false`, then the client knows it got an incomplete list, and will need to handle it.

Reply:
IMO, its a bit paranoic!

Reply:
I was thinking of a server-side graceful termination scenario, where the sidecar (or gRPC) had a chance to send an `io.EOF` before an unexpected termination, and not all the block metadata has been retrieved from the driver plugin. There is no way for the backup software to know the list it got is incomplete. Is there not a valid scenario?